webui/react/src/components/kit/CodeEditor.tsx

```diff
@@ -0,0 +1,329 @@
+
+    if (isConfig(fileInfo.key) || fileInfo.content !== NotLoaded) {
+      setActiveFile(fileInfo);
+      return;
+    }
+
+    let file,
+      content: Loadable<string> = NotLoaded;
+    try {
+      file = await fileInfo.get?.(String(fileInfo.key));
+    } catch (error) {
+      handleError(error, {
+        publicMessage: 'Failed to load selected file.',
+        publicSubject: 'Unable to fetch the selected file.',
+        silent: false,
+        type: ErrorType.Api,
+      });
+      setPageError(PageError.Fetch);
+    }
+    if (!file) {
+      setActiveFile({
+        ...fileInfo,
+        content: NotLoaded,
+      });
+      return;
+    }
+
+    try {
+      const text = decodeURIComponent(escape(window.atob(file)));
+
+      if (!text) setPageError(PageError.Empty); // Emmits a "Empty file" error message
+      content = Loaded(text);
+      setActiveFile({
+        ...fileInfo,
+        content,
+      });
+    } catch {
+      setPageError(PageError.Decode);
+    }
+  }, []);
+
+  useEffect(() => {
+    if (selectedFilePath && activeFile?.key !== selectedFilePath) {
+      const splitFilePath = selectedFilePath.split('/');
+      let matchTopFileOrFolder = null;
+      let fileDir = files;
+      for (let dir = 0; dir < splitFilePath.length; dir++) {
+        matchTopFileOrFolder = fileDir.find(
+          (f) => f.key === splitFilePath[dir] || f.key === selectedFilePath,
+        );
+        if (matchTopFileOrFolder?.children) {
+          fileDir = matchTopFileOrFolder.children;
+        }
+      }
+      if (matchTopFileOrFolder) {
+        fetchFile(matchTopFileOrFolder);
+      }
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  const handleSelectFile = useCallback(
+    (_: React.Key[], info: { node: TreeNode }) => {
+      const selectedKey = String(info.node.key);
+
+      if (selectedKey === activeFile?.key) {
+        // already selected
+        return;
+      }
+
+      const nodeAddress = selectedKey.split('/');
+
+      let targetNode = files.find((node) => node.title === nodeAddress[0]);
+      for (const dir of nodeAddress.slice(1))
+        targetNode = targetNode?.children?.find((file) => file.title === dir);
+
+      if (!targetNode) {
+        setPageError(PageError.Fetch);
+        return;
+      }
+
+      if (targetNode.isLeaf) {
+        onSelectFile?.(String(targetNode.key));
+        fetchFile(targetNode);
+      }
+    },
+    [activeFile?.key, fetchFile, files, onSelectFile],
+  );
+
+  const handleDownloadClick = useCallback(() => {
+    if (!activeFile) return;
+
+    const link = document.createElement('a');
+
+    link.download = isConfig(activeFile?.key)
+      ? activeFile.download || ''
+      : String(activeFile.title);
+    link.href = URL.createObjectURL(new Blob([Loadable.getOrElse('', activeFile?.content)]));
+    link.dispatchEvent(new MouseEvent('click'));
+    timeout.current = setTimeout(() => {
+      URL.revokeObjectURL(link.href);
+    }, 2000);
```

Comment:
does this still need to be stored in a ref?

Reply:
I don't think so, just was using that since that's how it was before. Removing it to just `setTimeout` now